lib/ddtrace/contrib/grpc/interception_context.rb

```diff
@@ -0,0 +1,47 @@
+require_relative 'datadog_interceptor'
+
+module GRPC
+  class InterceptionContext
+    # :nodoc:
+    # The `#intercept!` method is implemented in gRPC; this module
+    # will be prepended to the original class, effectively injecting
+    # our tracing middleware into the head of the call chain.
+    module InterceptWithDatadog
```

Comment:
Does this constant need to be defined in `GRPC::InterceptionContext`? (As opposed to the `Datadog::Contrib::GRPC` one?) We generally try to avoid opening/modifying classes or injecting new constants into integration namespaces, if possible.

Reply:
It does and is defined [here](https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/interceptors.rb#L154-L185). If you'd like, we can opt for the direct insertion of interceptors as in [this example](https://github.com/DataDog/dd-trace-rb/issues/379#issuecomment-381736597).

Reply:
I think I understand: for the default `use :grpc` scenario, we're not sure what RPC service to inject into, so we just prepend our module into the `InterceptionContext` yeah? If so, I think I should clarify what I meant; I think we can move `GRPC::InterceptionContext::InterceptWithDatadog` to `Datadog::Contrib::GRPC::InterceptWithDatadog` or something similar. The point of which is mostly to just avoid defining constants in the GRPC namespace, even if we do have to inject modules into constants defined there.

Reply:
Different question here: what happens if a user does `use :grpc` to have some default GRPC tracing, but then also adds:

```ruby
alternate_client = Demo::Echo::Service.rpc_stub_class.new(
  'localhost:50051',
  :this_channel_is_insecure,
  :interceptors => [configured_interceptor]
)
```

Would this cause this `alternate_client` to have two Datadog interceptors in it?

Reply:
That's a good question. The login in [`already_prepended?`](https://github.com/DataDog/dd-trace-rb/pull/403/files/1d8bfa857c40621b57ddf87b98fa70d9d09ec79c#diff-fed6c3884b710952ae65ea1663e6c3e1R32) should prevent that from happening. The `intercept!` call happens during the RPC interaction, so it should only add the default datadog interceptor to the call chain if the trace has not begun and if no other datadog interceptor has been supplied.